### PR TITLE
fix[flask]: ignore werkzeug.NotFound errors as server errors

### LINKED ISSUE #2347: werkzeug.exceptions.NotFound logged as error
Just recently started using ddtrace for APM, and perhaps I missed some configuration. If so, I apologize.

### Which version of dd-trace-py are you using?

0.46.0

### Which version of pip are you using?

21.0.1

### Which version of the libraries are you using?

datadog==0.41.0
ddtrace==0.46.0
Flask==1.1.2
Flask-RESTful==0.3.8
itsdangerous==1.1.0
Jinja2==2.11.3
MarkupSafe==1.1.1
Werkzeug==1.0.1

### How can we reproduce your problem?

Make a request to an endpoint that isn't handled by any Flask route.

### What is the result that you get?

An error logged to the trace for the `Flask.dispatch_request` operation like so:

```
werkzeug.exceptions.NotFound: 404 Not Found: The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.
Traceback (most recent call last):
File "/usr/lib/python3.8/site-packages/ddtrace/contrib/flask/patch.py", line 490, in _traced_request
return wrapped(*args, **kwargs)
File "/usr/lib/python3.8/site-packages/flask/app.py", line 1926, in dispatch_request
self.raise_routing_exception(req)
File "/usr/lib/python3.8/site-packages/flask/app.py", line 1908, in raise_routing_exception
raise request.routing_exception
File "/usr/lib/python3.8/site-packages/flask/ctx.py", line 350, in match_request
result = self.url_adapter.match(return_rule=True)
File "/usr/lib/python3.8/site-packages/werkzeug/routing.py", line 1945, in match
raise NotFound()
werkzeug.exceptions.NotFound: 404 Not Found: The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.
```

### What is the result that you expected?

For 404 requests not to result in a server error trace log. The `flask.request` operation is not logged as an error because at that point it knows the status code, but in `flask.dispatch_request` the `werkzeug.excpetions.NotFound` is raised and ddtrace thinks it is a server error.

Similarly, I would hope that other 4xx level werkzeug exceptions are also not logged as errors.

### PR DESCRIPTION
**Pull request by Yun-Kim**
## Description
It was found through #2347 that `werkzeug.exceptions.NotFound` 404 errors are currently raised and logged as server errors in `flask.dispatch_request` spans. This 404 error is no longer raised and logged as such.

Resolves #2347.

### CODE REVIEW

File: ddtrace/contrib/flask/patch.py [hunk 1]
```diff
@@ -463,29 +463,36 @@ def _traced_request(pin, wrapped, instance, args, kwargs):
 
         This wrapper will add identifier tags to the current span from `flask.app.Flask.wsgi_app`.
         """
-        span = pin.tracer.current_span()
-        if not pin.enabled or not span:
+        current_span = pin.tracer.current_span()
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/flask/patch.py:466]: Can we revert this variable name change? Makes the PR much bigger than it needs to be

File: ddtrace/contrib/flask/patch.py [hunk 2]
```diff
@@ -463,29 +463,36 @@ def _traced_request(pin, wrapped, instance, args, kwargs):
...
                 for k, v in request.view_args.items():
-                    span._set_str_tag(u".".join((FLASK_VIEW_ARGS, k)), v)
+                    current_span._set_str_tag(u".".join((FLASK_VIEW_ARGS, k)), v)
         except Exception:
             log.debug('failed to set tags for "flask.request" span', exc_info=True)
 
-        with pin.tracer.trace(".".join(("flask", name)), service=trace_utils.int_service(pin, config.flask, pin)):
+        with pin.tracer.trace(
+            ".".join(("flask", name)), service=trace_utils.int_service(pin, config.flask, pin)
+        ) as span:
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/flask/patch.py:494]:
```suggestion
        ) as request_span:
```

If the above rename was done to prevent clashing here

**Review by Kyle-Verhoog — COMMENTED** [review]
Just one concern - otherwise good to me!

File: tests/contrib/flask/test_errorhandler.py [hunk 3]
```diff
@@ -30,13 +30,10 @@ def test_default_404_handler(self):
         self.assertIsNone(req_span.get_tag("flask.url_rule"))
 
         # flask.dispatch_request span
-        self.assertEqual(dispatch_span.error, 1)
-        error_msg = dispatch_span.get_tag("error.msg")
-        self.assertTrue(error_msg.startswith("404 Not Found"))
-        error_stack = dispatch_span.get_tag("error.stack")
-        self.assertTrue(error_stack.startswith("Traceback (most recent call last):"))
-        error_type = dispatch_span.get_tag("error.type")
-        self.assertEqual(error_type, "werkzeug.exceptions.NotFound")
```
Comment (brettlangdon) [inline tests/contrib/flask/test_errorhandler.py:39]: oh nice, we already had a test case for this

**Review by brettlangdon — APPROVED** [review]

File: tests/contrib/flask/test_request.py [hunk 4]
```diff
@@ -529,10 +529,10 @@ def not_found():
...
         self.assertEqual(dispatch_span.name, "flask.dispatch_request")
         self.assertEqual(dispatch_span.resource, "flask.dispatch_request")
-        self.assertEqual(dispatch_span.error, 1)
-        self.assertTrue(dispatch_span.get_tag("error.msg").startswith("404 Not Found"))
-        self.assertTrue(dispatch_span.get_tag("error.stack").startswith("Traceback"))
-        self.assertEqual(dispatch_span.get_tag("error.type"), "werkzeug.exceptions.NotFound")
+        self.assertEqual(dispatch_span.error, 0)
+        self.assertIsNone(dispatch_span.get_tag("error.msg"))
+        self.assertIsNone(dispatch_span.get_tag("error.stack"))
+        self.assertIsNone(dispatch_span.get_tag("error.type"))
```
Comment (Yun-Kim) [inline tests/contrib/flask/test_request.py:535]: Note: the handler span propagates the error as well.

Comment (Yun-Kim) [inline tests/contrib/flask/test_request.py:535]: Note: the handler span propagates the error as well.

Comment (brettlangdon) [reply tests/contrib/flask/test_request.py:535]: which should be ok, as long as the main `flask.request` one doesn't have an exception attached, then it'll be ok

Comment (Kyle-Verhoog) [reply tests/contrib/flask/test_request.py:535]: @brettlangdon it still gets flagged as an error with exception information though which isn't great... 😕

Comment (brettlangdon) [reply tests/contrib/flask/test_request.py:535]: the `flask.request` one or the dispatch?

the dispatch one showing an exception should be ok as long as the main `flask.request` one doesn't? or am I missing something

Comment (Yun-Kim) [reply tests/contrib/flask/test_request.py:535]: So in normal cases of actual endpoints that don't exist, we get the expected behaviour (i.e. `flask.dispatch_request` span does not flag as error). However in the case of `test_404_abort`, if there is an actual endpoint but there's a manual `abort(404)` call, this creates a downstream handler span flagged with an error. Is this acceptable expected behaviour?

Comment (Kyle-Verhoog) [reply tests/contrib/flask/test_request.py:535]: Got it. I'm ok with it for now. Not great that the handler span will be flagged as erroneous but we can deal with that if requested.

**Review by Kyle-Verhoog — APPROVED** [review]

File: tests/contrib/flask/test_request.py (hunk 4 above)
Comment (brettlangdon) [reply tests/contrib/flask/test_request.py:535]: oh yeah, that behavior sounds good

**Pull request merged**